### ISSUE TITLE
Support ReflectionPad2d (and/or ConstantPad2d)

### ISSUE THREAD

**Issue by elboy** [onnx/onnx#21]
Running the command `torch_out = torch.onnx._export(torch_model, x, "saved_model.onnx", export_params=False)` for custom model architecture produces following error: 

Traceback (most recent call last):

  File "pytorch_to_caffe2.py", line 133, in <module>
    torch_out = torch.onnx._export(torch_model, x, "saved_model.onnx", export_params=False
  File "/opt/conda/lib/python2.7/site-packages/torch/onnx.py", line 54, in _export
    proto = trace.export(verbose)
  File "/opt/conda/lib/python2.7/site-packages/torch/nn/_function/thnn/auto.py", line 154, in symbolic
    return symbolic_fn(*args, **kwargs)
TypeError: 'NoneType' object is not callable

I went into auto.py and print out symbolic_fn and indeed it is None when I get this error. For tutorial (which worked fine) this function is `<function threshold_symoblic at 0x>`. 

When I print with verbose=True, I get the the following Error:

Error occured while handling:
%11 : Float(1, 3, 264, 264), %12 : Handle = ^ReflectionPad2d(4, 4, 4, 4)(%9), uses = [[%13.i0], []];

Exported graph so far:
graph(%1 : Float(32, 3, 9, 9)
        %2 : Float(32)
        %3 : Float(64, 32, 3, 3)
        %4 : Float(64)
        %5 : Float(32, 64, 3, 3)
        %6 : Float(32)
        %7 : Float(3, 32, 9, 9)
        %8 : Float(3)
        %9 : Float(1, 3, 256, 256)) {
  return ();

Any ideas of what might be going on and how to successfully export my pytorch model to caffe2? 

Thank you.

**Comment by ezyang** [comment]
`ReflectionPad2d` is not currenlty supported by the exporter, which is why you are seeing this error.

There is a legit bug; we are supposed to give a better error than `TypeError: 'NoneType' object is not callable`

**Comment by elboy** [comment]
Good to know, thank you. Any guess of the timeline for this support?

**Comment by ezyang** [comment]
Not sure, not sure that Caffe2 supports this op. We'll look into this. CC @houseroad 

By the way, I figured out why `TypeError: 'NoneType' object is not callable` is happening: in `auto.py` we're unconditionally creating a `symbolic` function, even when `symbolic_fn` is None. But the onnx C++ code relies on symbolic literally not being defined to determine if the operator is supported or not.

**Comment by colincsl** [comment]
@houseroad Any thoughts on this? This is blocking something of mine, so I would be happy to help if possible. Unfortunately I don't know where to look to fix this issue.

**Comment by houseroad** [comment]
I think right now, ONNX does not have pad related ops.  We are discussing introducing pad ops into ONNX.

**Comment by colincsl** [comment]
Thanks Lu. Do you have a rough timeline? e.g., days? weeks?

**Comment by ezyang** [comment]
Definitely not days, because it is not clear reflection padding can be implemented today with Caffe2; if this is indeed true, support for it would have to be added Caffe2 side first for an ONNX export to be useful.

**Comment by ebarsoum** [comment]
Wouldn't be more useful to have a more generic padding? Such as the on in tf.pad?

**Comment by ezyang** [comment]
Well, tf.pad isn't that generic; it just has a few modes, one of which corresponds to reflection padding. So it's mostly a matter of aesthetics whether or not it's a separate padding operator for each one, or a single padding operator with a string "type of padding."

**Comment by ezyang** [comment]
By the way, @colincsl, @bddppq has told me he is planning to implement reflection padding in Caffe2, so this op might be coming sooner than you might think :)

**Comment by colincsl** [comment]
Thanks @ezyang. While I wait on this, do you have any suggestions/trips for doing asymmetric padding? In particular this is for doing causal (temporal) convolutions. I don't need to be using reflection padding per se. 

As a reference, I have had an issue with assymetric padding with nn.ZeroPad2d, nn.ConstantPad2d, and when using the padding term in Conv2D (e.g., `self.clfs = nn.Conv2d(n_inputs, n_outputs, kernel_size=[1,1], padding=(0,0,2,0))`). This works fine when the padding term has 2 terms but breaks when it has 4 terms. There seems to be a `2*padding_tuple` somewhere, because it seems to be doubling the number of padding parameter. Perhaps this problem is actually more similar to this issue: https://github.com/pytorch/pytorch/issues/2752. Should I make a separate issue (on this repo) for that?

**Comment by dzhulgakov** [comment]
Actually Caffe2 supports PadImage with both constant and reflect: https://github.com/caffe2/caffe2/blob/master/caffe2/operators/pad_op.cc#L7 so adding the should be easy.

Anybody interested in cutting a PR?

As for aesthetics - I think one op with several modes is more concise, but I don't have strong preference.

**Comment by houseroad** [comment]
#73 Address the issue. But some work are needed on Pytorch and onnx-caffe2 side.

**Comment by houseroad** [comment]
The problem got solved, we can close this issue now.

**Comment by bddppq** [comment]
Thanks @houseroad !

**Comment by houseroad** [comment]
With https://github.com/pytorch/pytorch/pull/3048 , pytorch can export reflect/edge pad.
